ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	err = realWAL.StoreCheckpoint(writer, tries...)
+	if err != nil {
+		return fmt.Errorf("error serializing checkpoint (%d): %w", checkpointNum, err)
+	}
+
+	duration := time.Since(startTime)
+	logger.Info().Float64("total_time_s", duration.Seconds()).Msgf("created checkpoint %d", checkpointNum)
+
+	return nil
+}
+
+func cleanupCheckpoints(checkpointer *realWAL.Checkpointer, checkpointsToKeep int) error {
+	// Don't list checkpoints if we keep them all
+	if checkpointsToKeep == 0 {
+		return nil
+	}
+	checkpoints, err := checkpointer.Checkpoints()
+	if err != nil {
+		return fmt.Errorf("cannot list checkpoints: %w", err)
+	}
+	if len(checkpoints) > int(checkpointsToKeep) {
+		// if condition guarantees this never fails
+		checkpointsToRemove := checkpoints[:len(checkpoints)-int(checkpointsToKeep)]
+
+		for _, checkpoint := range checkpointsToRemove {
+			err := checkpointer.RemoveCheckpoint(checkpoint)
+			if err != nil {
+				return fmt.Errorf("cannot remove checkpoint %d: %w", checkpoint, err)
+			}
+		}
+	}
+	return nil
+}
+
+// processTrieUpdate writes trie update to WAL, updates activeSegmentNum,
+// and gets tries from checkpointQueue for checkpointing if needed.
+// It also sends WAL update result and waits for trie update completion.
+func (c *Compactor) processTrieUpdate(
+	update *WALTrieUpdate,
+	checkpointQueue *CheckpointQueue,
+	activeSegmentNum int,
+	nextCheckpointNum int,
+) (
+	_activeSegmentNum int,
+	checkpointNum int,
+	checkpointTries []*trie.MTrie,
+) {
+
+	// RecordUpdate returns the segment number the record was written to.
+	// Returned segment number (>= 0) can be
+	// - the same as previous segment number (same segment), or
+	// - incremented by 1 from previous segment number (new segment)
+	segmentNum, skipped, updateErr := c.wal.RecordUpdate(update.Update)
+
+	// This ensures that updated trie matches WAL update.
+	defer func(updateResult error) {
+		// Send result of WAL update
+		update.ResultCh <- updateResult
+
+		// Wait for updated trie
+		trie := <-update.TrieCh
+		if trie == nil {
+			c.logger.Error().Msg("compactor failed to get updated trie")
+			return
+		}
+
+		checkpointQueue.Push(trie)
+	}(updateErr)
+
+	if activeSegmentNum == -1 {
+		// Recover from failure to get active segment number at initialization.
+		return segmentNum, -1, nil
+	}
+
+	if updateErr != nil || skipped || segmentNum == activeSegmentNum {
+		return activeSegmentNum, -1, nil
+	}
+
+	// In the remaining code: segmentNum > activeSegmentNum
+
+	// active segment is finalized.
+
+	// Check new segment number is incremented by 1
+	if segmentNum != activeSegmentNum+1 {
+		c.logger.Error().Msg(fmt.Sprintf("compactor got unexpected new segment numer %d, want %d", segmentNum, activeSegmentNum+1))
+	}
+
+	// Update activeSegmentNum
+	prevSegmentNum := activeSegmentNum
+	activeSegmentNum = segmentNum
+
+	if nextCheckpointNum > prevSegmentNum {
```

Comment:
The logic of when to trigger checkpoint is spread in many places. I would suggest to create a CheckpointTrigger to decouple the logic from Compactor.

Here is the responsibility between them:
1. Compactor processes trieUpdate, forwards them to WAL writer. When WAL writer returns the segment num that the update was written to, it pass the segment num to CheckpointTrigger.
2. CheckpointTrigger is a stateful module, it holds the `activeSegmentNum`. When called with a segment num, it knows whether a new segment has been created, if created, it compares with its local checkpoint distance config to decide whether to trigger checkpoint. 
3. CheckpointTrigger holds the checkpointer module, when checkpoint is triggered, it places a lock and calls the checkpointer's method.
4. Checkpoint knows whether checkpointing has been running already, and can skip if the new checkpoint is triggered before the previous is finished
5. Compactor could pass the checkpointQueue to the checkpointTrigger along with the segment num, so that when checkpoint should be triggered, CheckpointTrigger has everything needed to trigger checkpoint.
6. Compactor knows nothing about checkpointing. Checkpoint related logic is all moved to the CheckpointTrigger, and can be tested much easier. 

```
type CheckpointTrigger struct {
  checkpointer // for running checkpointing
  checkpointDistance // the total number of segment (write-ahead log) to trigger checkpointing
  checkpointToKeep // the max number of checkpoint files to keep on disk in order to prevent too many checkpoint files filling disk.
  activeSegmentNum // the num of the segment that the previous trie update was trigger to, used to detect whether trie update is written to a new segment.
  isRunning // an atomic.Bool for whether checkpointing is currently running.
}

func (t *CheckpointTrigger) NotifyTrieUpdateWrittenToWAL(segmentNum int, checkpointQueue *CheckpointQueue) error {
  ...
}
```

Reply:
[Made this PR to explain the idea](https://github.com/onflow/flow-go/pull/2878/files)

Reply:
@zhangchiqing  I took a look at your PR and it looks like a refactor that doesn't fix a bug or blocker.  At a glance, it looks like there are some aspects that would require time for closer examination (the devil is in the details and we don't want to introduce bugs that are missed by existing tests).

Given the priority and urgency of this PR, I prefer refactorings of this size are done after we merge this PR.  🙏 

I think some of your suggestions are good and we should give them the time they deserve without rushing.

Reply:
sounds good.